### PR TITLE
Adding power to measure in uncertainty sampling for BQ

### PR DESCRIPTION
**Pull request by mmahsereci**
*Issue #377 

*Description of changes:*
- this is needed for the draft PR #375 
- added power to density
- adjusted the gradient
- some docstring fixes.

*tests*
- ran pytest and gradients seem to be correct still.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/quadrature/acquisitions/uncertainty_sampling.py [hunk 1]
```diff
@@ -11,63 +11,71 @@
...
-    The variance of the approximate transformed GP is used. If the integration measure is a probability measure,
-    then the variance will be weighted with the probability density at each point.
+    The acquisition function has the form :math:`a(x) = \var{x}` for the Lebesgue measure, and
+    :math:`a(x) = \var(x)p(x) ^ q` for a measure with density :math:`p(x)`. The default value for the power :math:`q`
+    is 2, but it can be set to a different value. :math:`\var(x)` is the posterior variance of the approximate
+    Gaussian process (GP) on the integrand.
     """
 
-    def __init__(self, model: Union[WarpedBayesianQuadratureModel, IDifferentiable]):
+    def __init__(self, model: Union[WarpedBayesianQuadratureModel, IDifferentiable], measure_power: float = None):
```
Comment (apaleyes) [inline emukit/quadrature/acquisitions/uncertainty_sampling.py:22]: Why not just?

```suggestion
    def __init__(self, model: Union[WarpedBayesianQuadratureModel, IDifferentiable], measure_power: float = 2):
```

File: emukit/quadrature/acquisitions/uncertainty_sampling.py [hunk 2]
```diff
@@ -11,63 +11,71 @@
...
         """
-        :param model: A warped Bayesian quadrature model that has predictive gradients
+        :param model: A warped Bayesian quadrature model that has gradients.
+        :param measure_power: The power of the measure. Default is 2. Only used if the measure is not the Lebesgue
+               measure.
         """
         self.model = model
+        self._measure_power = measure_power
+        if self._measure_power is None:
+            self._measure_power = 2.
```
Comment (apaleyes) [inline emukit/quadrature/acquisitions/uncertainty_sampling.py:31]: ... with suggestion above that's not needed

File: emukit/quadrature/acquisitions/uncertainty_sampling.py [hunk 3]
```diff
@@ -11,63 +11,71 @@
...
+
+        density = self.model.measure.compute_density(x)
+        density_gradient = self.model.measure.compute_density_gradient(x)
+
+        if p == 1:
             gradient_weighted = (density * variance_gradient.T).T + (variance[:, 0] * density_gradient.T).T
             return variance_weighted, gradient_weighted
+
+        gradient_weighted = (density ** p * variance_gradient.T
+                             + p * (variance[:, 0] * density ** (p - 1)) * density_gradient.T).T
```
Comment (apaleyes) [inline emukit/quadrature/acquisitions/uncertainty_sampling.py:79]: I think that would benefit from other formatting, or braces, or anything breaking it into smaller chunks. It's hard for me to immediately figure out the order of evaluation of all these operators.

File: emukit/quadrature/acquisitions/uncertainty_sampling.py [hunk 4]
```diff
@@ -11,63 +11,71 @@
...
+    is 2, but it can be set to a different value. :math:`\var(x)` is the posterior variance of the approximate
+    Gaussian process (GP) on the integrand.
     """
 
-    def __init__(self, model: Union[WarpedBayesianQuadratureModel, IDifferentiable]):
+    def __init__(self, model: Union[WarpedBayesianQuadratureModel, IDifferentiable], measure_power: float = None):
         """
-        :param model: A warped Bayesian quadrature model that has predictive gradients
+        :param model: A warped Bayesian quadrature model that has gradients.
+        :param measure_power: The power of the measure. Default is 2. Only used if the measure is not the Lebesgue
```
Comment (apaleyes) [inline emukit/quadrature/acquisitions/uncertainty_sampling.py:25]: Does the power have any restrictions? E.g. can it be negative?

**Review by apaleyes — COMMENTED** [review]
Looks good, just a few minor comments.

File: emukit/quadrature/acquisitions/uncertainty_sampling.py (hunk 2 above)
Comment (mmahsereci) [reply emukit/quadrature/acquisitions/uncertainty_sampling.py:31]: done

File: emukit/quadrature/acquisitions/uncertainty_sampling.py (hunk 3 above)
Comment (mmahsereci) [reply emukit/quadrature/acquisitions/uncertainty_sampling.py:79]: I put braces around each term in the sum now. Hope this improves readability.

File: emukit/quadrature/acquisitions/uncertainty_sampling.py (hunk 4 above)
Comment (mmahsereci) [reply emukit/quadrature/acquisitions/uncertainty_sampling.py:25]: I don;t think is has restrictions if density is positive (which it should be). Although I am not sure what kind of uncertainty sampling that would give if the variance was divided by the density. In practice mostly p=2 and p=1 are used.

**Pull request merged**